external/js/cothority/src/byzcoin/v2/darcsBS.ts

```diff
@@ -0,0 +1,268 @@
+        );
+        const bsDarc = await ObservableToBS(instObs);
+        if (bsDarc.getValue() === undefined) {
+            throw new Error("this darc doesn't exist");
+        }
+        return new DarcBS(bsDarc);
+    }
+
+    constructor(darc: BehaviorSubject<DarcInst>) {
+        super(darc.getValue());
+        darc.subscribe(this);
+    }
+
+    /**
+     * Creates an instruction in the transaction with either an update of the description and/or an update
+     * of the rules.
+     *
+     * @param tx where the instruction will be appended to
+     * @param updates contains a description update and/or rules to be merged
+     * @param unrestricted if true, will create an unrestricted evolve that allows to create new rules
+     * @return the new DARC as it will appear on ByzCoin if the transaction is accepted
+     */
+    evolve(tx: TransactionBuilder, updates: IDarcAttr, unrestricted = false): Darc {
+        return DarcInst.evolve(tx, this.getValue(), updates, unrestricted);
+    }
+
+    /**
+     * Sets the description of the DARC.
+     *
+     * @param tx where the instruction will be appended to
+     * @param description of the new DARC
+     * @return the new DARC as it will appear on ByzCoin if the transaction is accepted
+     */
+    setDescription(tx: TransactionBuilder, description: Buffer): Darc {
+        return this.evolve(tx, {description});
+    }
+
+    /**
+     * Sets the sign and evolve rules of the darc.
+     *
+     * @param tx where the instruction will be appended to
+     * @param idSign the IIdentity or the InstanceID of the DARC that is delegated
+     * @param idEvolve per default the same as idSign, but can be given as a different argument
+     * @return the new DARC as it will appear on ByzCoin if the transaction is accepted
+     */
+    setSignEvolve(tx: TransactionBuilder, idSign: IIdentity | InstanceID, idEvolve = idSign): Darc {
+        const rules = this.getValue().rules.clone();
+        rules.setRule(Darc.ruleSign, toIId(idSign));
+        if (idEvolve) {
+            rules.setRule(DarcInstance.ruleEvolve, toIId(idEvolve));
+        }
+        return this.evolve(tx, {rules});
+    }
+
+    /**
+     * Adds a sign and evolve element to the DARC with an OR expression
+     *
+     * @param tx where the instruction will be appended to
+     * @param idSign IIdentity or DARC-ID to be added to the signature rule. If it is undefined, nothing will be
+     * added to the signature rule
+     * @param idEvolve IIdentity or DARC-ID to be added to the evolve rule. Per default the same as the idSign.
+     * @return the new DARC as it will appear on ByzCoin if the transaction is accepted
+     */
+    addSignEvolve(tx: TransactionBuilder, idSign: IIdentity | InstanceID | undefined, idEvolve = idSign): Darc {
+        const rules = this.getValue().rules.clone();
+        if (idSign) {
+            rules.appendToRule(Darc.ruleSign, toIId(idSign), Rule.OR);
+        }
+        if (idEvolve) {
```

Comment:
is always != undefined

Reply:
no, you can pass `undefined` in the method.

Reply:
yes, you can pass, but as @tharvik writes, ts decides that `undefined` === no argument...